examples/deepspeed-multinode/sky.yaml

```diff
@@ -0,0 +1,80 @@
+# Example: a distributed DeepSpeed job (DeepSpeed-Chat) on 2 VMs.
+#
+# This takes care constructing a "hostfile" to pass to DeepSpeed.
+#
+# Usage:
+#
+#   $ sky launch sky.yaml -r --down -c ds
+#
+#   # After the job starts running, log into the two nodes and check gpustat:
+#
+#   $ ssh ds
+#   $ gpustat -i
+#
+#   $ ssh ds-worker1
+#   $ gpustat -i
+
+resources:
+  accelerators: T4:1  # AWS, Azure, GCP
+  # accelerators: A100:1  # GCP, Lambda
+  # accelerators: A100-80GB:1  # Azure, GCP, SCP
+num_nodes: 2
+
+setup: |
+  set -ex
+  git clone https://github.com/microsoft/DeepSpeedExamples.git || true
+  cd DeepSpeedExamples
+  git checkout d7c42b4f34df91035e7ed3e0c51500bb53d0bc71
+
+  conda create -n deepspeed python=3.8 -y
+  source activate deepspeed
+
+  pip install deepspeed
+
+  cd applications/DeepSpeed-Chat
+  pip install -r requirements.txt
+
+  # Required by DeepSpeed in multi-node settings.
+  sudo apt-get -y install pdsh
+
+file_mounts:
+  # Required for DeepSpeed's passwordless SSH (run commands on nodes).
+  ~/.ssh/id_rsa: ~/.ssh/sky-key
+
+run: |
+  cd DeepSpeedExamples
+  source activate deepspeed
```

Comment:
```suggestion
  conda activate deepspeed
```

Reply:
Oddly, on GCP, this doesn't work:
```
(head, rank=0, pid=23677) 10.182.0.52:   File "/opt/conda/envs/deepspeed/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1649, in verify_ninja_availability
(head, rank=0, pid=23677) 10.182.0.52:     raise RuntimeError("Ninja is required to load C++ extensions")
(head, rank=0, pid=23677) 10.182.0.52: RuntimeError: Ninja is required to load C++ extensions
```
due to the activation failing at task exec time, and the `base` or system env is used instead:
```
(base) gcpuser@ray-ds2-head-aaa1e0e7-compute:~$ pip3 list | grep ninja
(base) gcpuser@ray-ds2-head-aaa1e0e7-compute:~$ conda activate deepspeed
(deepspeed) gcpuser@ray-ds2-head-aaa1e0e7-compute:~$ pip3 list | grep ninja
ninja                    1.11.1
```
With `source activate` it works.

Reply:
Hmm, that is weird. This seems to be a serious issue. We may need to look into.

Reply:
Actually, I'm running a really weird issue.
```
sky launch sky.yaml -r -i60 -c ds --cloud gcp
```
failed with
```
..
(head, rank=0, pid=22066) 10.142.15.204: RuntimeError: Ninja is required to load C++ extensions
(head, rank=0, pid=22066) 10.142.15.203: RuntimeError: Ninja is required to load C++ extensions
```
which [meant](https://github.com/pytorch/pytorch/blob/main/torch/utils/cpp_extension.py#L1645) `ninja --version` failed.

However, after ssh into the cluster:
```
gcpuser@ray-ds-head-2500deff-compute:~$ source activate deepspeed
(deepspeed) gcpuser@ray-ds-head-2500deff-compute:~$ ninja --version
1.11.1.git.kitware.jobserver-1
```
So it's installed in the conda env. 

Manually running these failed with the same error
- `python .sky/sky_app/sky_job_1`
- (printing out the innermost command we pass to Popen()) `/bin/bash -i /tmp/sky_app_96ypzgks`

However, manually running just the activate + the deepspeed launch command works
```
gcpuser@ray-ds-head-2500deff-compute:~$ source activate deepspeed
(deepspeed) gcpuser@ray-ds-head-2500deff-compute:~$ deepspeed \
>       main.py \
>       --data_path Dahoas/rm-static Dahoas/full-hh-rlhf Dahoas/synthetic-instruct-gptj-pairwise yitingxie/rlhf-reward-datasets \
>       --data_split 2,4,4 \
>       --model_name_or_path facebook/opt-1.3b \
>       --per_device_train_batch_size 8 \
>       --per_device_eval_batch_size 8 \
>       --max_seq_len 512 \
>       --learning_rate 1e-3 \
>       --weight_decay 0.1 \
>       --num_train_epochs 16 \
>       --gradient_accumulation_steps 1 \
>       --lr_scheduler_type cosine \
>       --num_warmup_steps 0 \
>       --seed 1234 \
>       --zero_stage 0 \
>       --lora_dim 128 \
>       --lora_module_name decoder.layers. \
>       --only_optimize_lora \
>       --deepspeed \
>       --output_dir $OUTPUT_PATH
```
training proceeds as normal.

Any help?